pkg/scanfetcher/nexpose.go

```diff
@@ -0,0 +1,172 @@
+package scanfetcher
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/asecurityteam/nexpose-scan-notifier/pkg/domain"
+)
+
+const (
+	activeQueryParam = "active" // Bool indicating if active scans should be fetched.
+	pageQueryParam   = "page"   // The index of the page (zero-based) to retrieve.
+	sizeQueryParam   = "size"   // The number of records per page to retrieve.
+	sortQueryParam   = "sort"
+	sortQueryValue   = "endTime,DESC" // Return scans in descending order start with most recently completed.
+
+	finishedScanStatus = "finished" // Status for scans which have completed successfully.
+)
+
+type page struct {
+	Number         int `json:"number"`
+	Size           int `json:"size"`
+	TotalPages     int `json:"totalPages"`
+	TotalResources int `json:"totalResources"`
+}
+
+type resource struct {
+	EndTime string `json:"endTime"`
+	ScanID  int    `json:"id"`
+	SiteID  int    `json:"siteId"`
+	Status  string `json:"status"`
+}
+
+type nexposeScanResponse struct {
+	Page      page       `json:"page"`
+	Resources []resource `json:"resources"`
+}
+
+// NexposeClient implements the interfaces to fetch scans from Nexpose
+type NexposeClient struct {
+	Client   *http.Client
+	Host     *url.URL
+	PageSize int
+}
+
+// FetchScans fetches Nexpose scans, filters out running scans, and returns all completed scans
+// after the provided timestamp.
+func (n *NexposeClient) FetchScans(ctx context.Context, ts time.Time) ([]domain.CompletedScan, error) {
+	var completedScans []domain.CompletedScan
+
+	scanResp, err := n.makePagedNexposeScanRequest(0)
+	if err != nil {
+		return nil, err
+	}
+
+	pages := scanResp.Page.TotalPages
+	for _, resource := range scanResp.Resources {
+		completedScan, err := scanResourceToCompletedScan(resource, ts)
+		switch err.(type) {
+		case nil:
+			completedScans = append(completedScans, completedScan)
+		case scanNotFinishedError:
+			// skip scans without a status of "finished"
+		case outOfRangeError:
+			// since scans are returned in descending order by scan time, return
+			// the list of completed scans after finding the first scan outside
+			// the valid time range
+			return completedScans, nil
+		default:
+			return nil, err
+		}
+	}
+
+	for curPage := 1; curPage < pages; curPage = curPage + 1 {
+		scanResp, err := n.makePagedNexposeScanRequest(curPage)
```

Comment:
question: would this be a candidate for using a fan-out? or would nexpose not be able to handle that?

Reply:
I'm not sure about the Nexpose performance implications, but in the general case I wouldn't think we'd really ever need to fetch more than one page of scans. We're only looking for the most recent scans, so once we process a scan we should never need to fetch the results again.

Reply:
Ah i missed the return case below. carry on